#!/bin/bash 

"""Run taxonomic identifier Kraken 2 on fastq files generated by Oxford Nanopore sequencing, then estimate the abundance of different species in the sample."""

__appname__ = 'kraken2_run.sh'
__author__ = 'Viki Blanchard (viki.blanchard.2018@live.rhul.ac.uk)'
__version__ = '0.0.1'


""" GET ISOLATE NAME """ 

#######################
### Set up Kraken 2 ###
#######################

# Activate conda environment to run the analysis in and install kraken, kraken 2, and bracken into the environment

conda create --yes -n kraken -c bioconda kraken2 bracken
conda activate kraken

# Navigate to analysis root directory
mkdir /media/vlb19/Expansion/analysis
cd /media/vlb19/Expansion/analysis

# Create Kraken analysis directory and navigate into it
mkdir kraken
cd kraken

# Download Kraken standard database 
## NB: requires ~450GB of additional disk space and 29GB of RAM
## 24 threads on a computer with 252GB of RAM takes 1.5 hours

"""Use MiniKraken for practice run"""
# Unzip database
tar -xf *.tar.gz

##############################################################
### Build a custom Kraken database for likely contaminants ###
##############################################################

### Steps from Kraken User Guide
# 1.) Install a taxonomy
#       Usually you can just use the NCBI taxonomy, which you can download using: kraken-build --download-taxonomy --db $DBNAME 

""" Bombus NCBI Database Taxonomy ID: 144708 """

################################################################
### Create a report for taxonomy of sequences in fastq files ###
################################################################

kraken2 --use-names --threads 4 --db """PATH_TO_DB_DIR""" --report ${isolate} --fastq-input Results/${isolate}_high_qual_reads.fastq.gz > ${isolate}.kraken

################################################################
### Estimate species and genus level abundances with Bracken ###
################################################################

# generate report on species-level abundances
bracken -d """PATH_TO_DB_DIR""" -i ${isolate}.kraken -o ${isolate}_species.bracken -l S 

# generate report on genus-level abundances
bracken -d """PATH_TO_DB_DIR""" -i ${isolate}.kraken -o ${isolate}_genus.bracken -l G

# generate report on family-level abundances
bracken -d """PATH_TO_DB_DIR""" -i ${isolate}.kraken -o ${isolate}_genus.bracken -l F